nomad/job_endpoint.go

```diff
@@ -1139,14 +1139,38 @@ func (j *Job) GetJobVersions(args *structs.JobVersionsRequest,
 	return j.srv.blockingRPC(&opts)
 }
 
+func (j *Job) validateListAllJobsPermission(aclObj *acl.ACL, state *state.StateStore) error {
+	if aclObj == nil || aclObj.IsManagement() {
+		return nil
+	}
+
+	// namespaces
+	nses, err := state.NamespaceNames()
+	if err != nil {
+		return err
+	}
+
+	for _, ns := range nses {
+		if !aclObj.AllowNsOp(ns, acl.NamespaceCapabilityListJobs) {
+			return structs.ErrPermissionDenied
```

Comment:
since this is already linear-time/memory on the number of namespaces + the number of jobs, why not use the token as a filter?

Reply:
Not sure I follow the intent - is the suggestion to keep the filtering logic but call `aclObj.AllowNsOps(...)` on every job rather than every namespaces?  That's seems unnecessary wasteful.  I'd assume there are orders of magnitudes more jobs than namespaces.

Or is the suggestion to move it to job level but then only show the jobs that the acl is allowed to view.  This is reasonable, but I'd be confused if I pass `all_namespaces=true` but got only a subset.

Reply:
The latter.

> but I'd be confused if I pass all_namespaces=true but got only a subset.

That's exactly the behavior of `GET /namespaces`; the list of namespaces is filtered by ACL.

Reply:
Good point.  I'll update it.

Reply:
I'll go further and argue that filtering is more consistent. It means that `GET /jobs?all_namespaces=true` is equivalent to `GET /jobs | filter(_.Namespace in GET /namespaces)`. 

Without a filtering approach, users have the situation where `GET /namespaces` returns a list of namespaces and `GET /job?all_namespaces=true` returns 403, so that they're required to loop over `GET /jobs` for all namespaces returns by `GET /namespaces`; this seems to me the exact situation that this modification is intended to avoid.

Reply:
(answered at the same time, didn't mean to pile on)